lib/server.js

```diff
@@ -0,0 +1,234 @@
+      const format = match[1]
+      const badgeData = makeBadgeData('404', query)
+      badgeData.text[1] = 'badge not found'
+      badgeData.colorscheme = 'red'
+      // Add format to badge data.
+      badgeData.format = format
+      const svg = makeBadge(badgeData)
+      makeSend(format, request.res, end)(svg)
+    })
+
+    camp.notfound(/.*/, (query, match, end, request) => {
+      end(null, { template: '404.html' })
+    })
+  }
+
+  registerServices() {
+    const { config, camp } = this
+    const { githubApiProvider } = this.githubConstellation
+
+    loadServiceClasses().forEach(serviceClass =>
+      serviceClass.register(
+        { camp, handleRequest, githubApiProvider },
+        {
+          handleInternalErrors: config.handleInternalErrors,
+          cacheHeaders: config.cacheHeaders,
+          profiling: config.profiling,
+        }
+      )
+    )
+  }
+
+  registerRedirects() {
+    const { config, camp } = this
+
+    // Any badge, old version. This route must be registered last.
+    camp.route(/^\/([^/]+)\/(.+).png$/, (queryParams, match, end, ask) => {
+      const [, label, message] = match
+      const { color } = queryParams
+
+      const redirectUrl = staticBadgeUrl({
+        label,
+        message,
+        color,
+        format: 'png',
+      })
+
+      ask.res.statusCode = 301
+      ask.res.setHeader('Location', redirectUrl)
+
+      // The redirect is permanent.
+      const cacheDuration = (365 * 24 * 3600) | 0 // 1 year
+      ask.res.setHeader('Cache-Control', `max-age=${cacheDuration}`)
+
+      ask.res.end()
+    })
+
+    if (config.redirectUri) {
+      camp.route(/^\/$/, (data, match, end, ask) => {
+        ask.res.statusCode = 302
+        ask.res.setHeader('Location', config.redirectUri)
+        ask.res.end()
+      })
+    }
+  }
+
+  async start() {
+    const {
+      bind: { port, address: hostname },
+      ssl: { isSecure: secure, cert, key },
+      cors: { allowedOrigin },
+      baseUri,
+      rateLimit,
+    } = this.config
+
+    log(`Server is starting up: ${baseUri}`)
+
+    const camp = (this.camp = Camp.start({
+      documentRoot: path.join(__dirname, '..', 'public'),
+      port,
+      hostname,
+      secure,
+      cert,
+      key,
+    }))
+
+    analytics.load()
+    analytics.scheduleAutosaving()
+    analytics.setRoutes(camp)
+
+    this.cleanupMonitor = sysMonitor.setRoutes({ rateLimit }, camp)
+
+    const { githubConstellation, metrics } = this
+    githubConstellation.initialize(camp)
+    metrics.initialize(camp)
+
+    const { githubApiProvider } = this.githubConstellation
+    suggest.setRoutes(allowedOrigin, githubApiProvider, camp)
+
+    this.registerErrorHandlers()
+    this.registerServices()
+
+    await new Promise(resolve => camp.on('listening', () => resolve()))
+  }
+
+  static resetGlobalState() {
+    clearRequestCache()
+    clearRegularUpdateCache()
+  }
+
+  reset() {
```

Comment:
Why do we need both `resetGlobalState()` and `reset()` here?

Reply:
Over time we should transition this stuff from global state to instance state.